File: app/src/test/java/com/awscherb/cardkeeper/ui/cards/CardsPresenterTest.java
package com.awscherb.cardkeeper.ui.cards;

import com.awscherb.cardkeeper.data.model.ScannedCode;
import com.awscherb.cardkeeper.data.service.ScannedCodeService;

import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import java.util.ArrayList;

import io.reactivex.Observable;

import static org.mockito.Matchers.any;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class CardsPresenterTest {

    @Mock private ScannedCodeService scannedCodeService;
    @Mock private CardsContract.View view;
    private CardsContract.Presenter presenter;

    @Before
    public void setup() {
        MockitoAnnotations.initMocks(this);

        presenter = new CardsPresenter(view, scannedCodeService);
    }

    @Test
    public void test_loadCodes() {
        when(scannedCodeService.listAllScannedCodes()).thenReturn(Observable.just(new ArrayList<>()));

        presenter.loadCards();

        verify(view).showCards(any());
    }

    @Test
    public void test_addCode() {
        when(scannedCodeService.addScannedCode(any(ScannedCode.class))).thenReturn(Observable.just(new ScannedCode()));

        presenter.addNewCard(any(ScannedCode.class));

        verify(view).onCardAdded(any(ScannedCode.class));
    }

    @Test
    public void test_deleteCode() {
        when(scannedCodeService.deleteScannedCode(any(ScannedCode.class))).thenReturn(Observable.just(null));

        presenter.deleteCard(any(ScannedCode.class));

        verify(view).onCardDeleted();
    }

}

File: app/src/test/java/com/awscherb/cardkeeper/ui/cards/CardsPresenterTest.kt
package com.awscherb.cardkeeper.ui.cards

import com.awscherb.cardkeeper.data.model.ScannedCode
import com.awscherb.cardkeeper.data.service.ScannedCodeService
import com.awscherb.cardkeeper.ui.base.BasePresenterTest
import com.nhaarman.mockitokotlin2.any
import com.nhaarman.mockitokotlin2.times
import com.nhaarman.mockitokotlin2.verify
import com.nhaarman.mockitokotlin2.whenever
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.runBlocking
import org.junit.Test
import org.mockito.Mock

class CardsPresenterTest : BasePresenterTest<CardsContract.View, CardsPresenter>() {

    @Mock
    lateinit var service: ScannedCodeService

    override fun getViewClass() = CardsContract.View::class.java

    override fun createPresenter() = CardsPresenter(Dispatchers.Unconfined, service)

    @Test
    fun `Load cards success`() = runBlocking {
        whenever(service.listAllScannedCodes())
            .thenReturn(ArrayList())

        presenter.loadCards()

        verify(view, times(1)).showCards(any())
    }

    @Test
    fun `Load cards failure`() = runBlocking {
        whenever(service.listAllScannedCodes())
            .thenThrow(RuntimeException())

        presenter.loadCards()

        verify(view, times(1)).onError(any())
    }

    @Test
    fun `Add new card success`() = runBlocking {
        val newCode = ScannedCode()
        whenever(service.addScannedCode(newCode))
            .thenReturn(newCode)

        presenter.addNewCard(newCode)

        verify(view, times(1)).onCardAdded(newCode)
    }

    @Test
    fun `Add new card failure`() = runBlocking {
        whenever(service.addScannedCode(any()))
            .thenThrow(RuntimeException())

        presenter.addNewCard(ScannedCode())

        verify(view, times(1)).onError(any())
    }

    @Test
    fun `Delete card success`() = runBlocking {
        //        whenever(service.deleteScannedCode(any()))
//            .thenAnswer {  }

        presenter.deleteCard(ScannedCode())

        verify(view, times(1)).onCardDeleted()
    }

    @Test
    fun `Delete card failure`() = runBlocking {
        whenever(service.deleteScannedCode(any()))
            .thenThrow(RuntimeException())

        presenter.deleteCard(ScannedCode())

        verify(view, times(1)).onError(any())
    }
}

